### PR TITLE
arm64/a64: add driver for Allwinner A64 I2C bus

### PR DESCRIPTION
**Pull request by zouboan**
## Summary
1. arm64/a64: add driver for Allwinner A64 I2C bus(TWI controller);
2. boards/pinephone: add support of on board MPU6050 sensor;
3. fixed some build warning in aarch64 gcc which will result error in ci
## Impact
a64 and pinephone
## Testing
MPU6050

### CODE REVIEW

**Comment by xiaoxiang781216** [comment]
@lupyuen have you associate your github account with your Apache ID? I found I can't add you to reviewer list.

**Comment by lupyuen** [comment]
@xiaoxiang781216 Yes my GitHub ID is already in my Whimsy Profile. Is there anything else I need to set?

Lemme review this PR thanks!

**Comment by zouboan** [comment]
@xiaoxiang781216 @lupyuen Sorry i found an issue, please don't review this PR until i fixed it!

**Comment by lupyuen** [comment]
Sure @zouboan :-) I have a suggestion about the addresses: in a64_memorymap.h we're proposing:

```c
#define A64_RTWI_VBASE      0x01f02400 /* Base address of R_TWI */
#define A64_TWI0_VBASE      0x01c2ac00 /* Base address of TWI0 */
#define A64_TWI1_VBASE      0x01c2b000 /* Base address of TWI1 */
#define A64_TWI2_VBASE      0x01c2b400 /* Base address of TWI2 */
```

May I suggest that we sort these addresses together with the previous section, for easier maintenance?

```c
#define A64_DE_ADDR         0x01000000 /* DE              0x0100:0000-0x012f:ffff 3M */
#define A64_SYSCTL_ADDR     0x01c00000 /* System Control  0x01c0:0000-0x01c0:0fff 4K */
...
```

Also could we rename `VBASE` to `ADDR` to be consistent?

BTW: I derived the addresses from the Allwinner A10 Memory Map in NuttX:

https://github.com/apache/nuttx/blob/master/arch/arm/src/a1x/hardware/a10_memorymap.h#L60-L127

The rightmost column looks similar to the Memory Mapping (Page 73) of the A64 User Manual. Thanks :-)

**Comment by zouboan** [comment]
@lupyuen Thanks very much for your advise, i will change it

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216 Yes my GitHub ID is already in my Whimsy Profile. Is there anything else I need to set?
> 
> Lemme review this PR thanks!

can you add yourself to reviewer at the top right?

**Comment by lupyuen** [comment]
@xiaoxiang781216 Sorry GitHub won't let me add as reviewer...

Screenshot 2022-12-30 at 9 41 56 PM

**Comment by xiaoxiang781216** [comment]
So, look like your account setup does not finish yet. @btashton @hartmannathan and @pkarashchenko do you remember the step to get the github review and merge permission?

**Comment by pkarashchenko** [comment]
The Apache account setup has to be complete Apache GitHub organisation should confirm GitHub account linked to Apache member account. I will try to search for details, but it took me a week after I filled all the data before my GitHub account was linked.

**Comment by hartmannathan** [comment]
> So, look like your account setup does not finish yet. @btashton @hartmannathan and @pkarashchenko do you remember the step to get the github review and merge permission?

I know you need to have 2 factor authentication and that if not done within some timeout since the invite then problems ensue. Maybe ask to Infra at Apache?

**Comment by zouboan** [comment]
@lupyuen The issue was fixed, please review this PR, thanks!

**Comment by lupyuen** [comment]
@zouboan Thanks! I'm reviewing now. In a64_twi.c:

```c
#define BUS_CLK_GATING_REG3 (A64_CCU_ADDR + 0x006C)
```

Let's change `6C` to `6c` to be consistent.

**Comment by lupyuen** [comment]
In a64_twi.c: Let's insert a blank line after `#endif`

```c
#endif
/****************************************************************************
 * Private Functions
```

**Comment by lupyuen** [comment]
In a64_twi.c: Let's insert a space before `Put`:

```c
 * Description:
 *  Put a 32-bit register value by offset
```

Let's insert 2 spaces before `modreg32` and `putreg32`:

```c
 * Description:
 *   Modify a 32-bit register value by offset
 * modreg32(v,m,a) defined as:
 * putreg32((getreg32(a) & ~(m)) | ((v) & (m)), (a))
```

Could we do the same alignment for the other `Description` sections?

**Comment by lupyuen** [comment]
In a64_memorymap.h:

```c
#define A64_RTWI_ADDR       0x01f02400 /* R_TWI           0x01f0:2400-0x01f0:27ff 1K */
#define A64_TWI0_ADDR       0x01c2ac00 /* TWI 0           0x01c2:ac00-0x01c2:afff 1K */
#define A64_TWI1_ADDR       0x01c2b000 /* TWI 1           0x01c2:b000-0x01c2:B3ff 1K */
#define A64_TWI2_ADDR       0x01c2b400 /* TWI 2           0x01c2:b400-0x01c2:b7ff 1K */
```

Let's sort together with the other addresses, like this:

```c
#define A64_DE_ADDR         0x01000000 /* DE              0x0100:0000-0x012f:ffff 3M */
#define A64_SYSCTL_ADDR     0x01c00000 /* System Control  0x01c0:0000-0x01c0:0fff 4K */
#define A64_TCON0_ADDR      0x01c0c000 /* TCON 0          0x01c0:c000-0x01c0:cfff 4K */
#define A64_CCU_ADDR        0x01c20000 /* CCU             0x01c2:0000-0x01c2:03ff 1K */
#define A64_PIO_ADDR        0x01c20800 /* PIO             0x01c2:0800-0x01c2:0bff 1K */
#define A64_PWM_ADDR        0x01c21400 /* PWM             0x01c2:1400-0x01c2:17ff 1K */
#define A64_TWI0_ADDR       0x01c2ac00 /* TWI 0           0x01c2:ac00-0x01c2:afff 1K */
#define A64_TWI1_ADDR       0x01c2b000 /* TWI 1           0x01c2:b000-0x01c2:B3ff 1K */
#define A64_TWI2_ADDR       0x01c2b400 /* TWI 2           0x01c2:b400-0x01c2:b7ff 1K */
#define A64_DSI_ADDR        0x01ca0000 /* MIPI DSI        0x01ca:0000-0x01ca:0fff 4K */
#define A64_DPHY_ADDR       0x01ca1000 /* MIPI DSI-PHY    0x01ca:1000-0x01ca:1fff 4K */
#define A64_RTWI_ADDR       0x01f02400 /* R_TWI           0x01f0:2400-0x01f0:27ff 1K */
#define A64_RPIO_ADDR       0x01f02c00 /* R_PIO           0x01f0:2c00-0x01f0:2fff 1K */
#define A64_RSB_ADDR        0x01f03400 /* R_RSB           0x01f0:3400-0x01f0:37ff 1K */
#define A64_RPWM_ADDR       0x01f03800 /* R_PWM           0x01f0:3800-0x01f0-3bff 1K */
```

**Comment by lupyuen** [comment]
In a64_twi.h:

```c
#define A64_TWI_CNTR_OFFSET         (0x0C)  /* 31:8 bit reserved, INT_EN, BUS_EN, M_STA, INT_FLAG, A_ACK */
#define A64_TWI_EFR_OFFSET     	    (0x1C) 	/* 31:2 bit reserved,1:0 bit data byte follow read comand */
(And other lines)
```

Let's standardise on lower case for the hex numbers in the file.

**Comment by lupyuen** [comment]
In a64_twi.h: Let's align the numbers:

```c
#define TWI_STAT_ARBLOST_RXGCAS_ACK (0x78)
#define TWI_STAT_RXDS_ACK	          (0x80)
#define TWI_STAT_RXDS_NAK	          (0x88)
#define TWI_STAT_RXDGCAS_ACK        (0x90)
```

**Comment by lupyuen** [comment]
In a64_twi.c:

```c
  switch (status)
    {
    case 0xf8: /* On reset or stop the bus is idle, used only on polled */
      err_code = 0xf8;
      goto errout;
    case 0x08: /* A START condition has been transmitted */
    case 0x10: /* A repeated start condition has been transmitted */
```

The [coding standard](https://nuttx.apache.org/docs/latest/contributing/coding_style.html#switch-statement) says we should insert 2 spaces before `case`, like this:

```c
  switch (status)
    {
      case 0xf8: /* On reset or stop the bus is idle, used only on polled */
        err_code = 0xf8;
        goto errout;
      case 0x08: /* A START condition has been transmitted */
      case 0x10: /* A repeated start condition has been transmitted */
```

**Comment by lupyuen** [comment]
The I2C logic looks OK when I checked with the A64 User Manual. So my review is complete.

After the I2C driver has been merged, I look forward to testing your driver with the [PinePhone I2C Touch Panel](https://lupyuen.github.io/articles/pio#touch-panel). Thank you so much! :-)

**Comment by lupyuen** [comment]
Sorry one last comment: In pinephone_bringup.c:

```c
#include <nuttx/sensors/mpu60x0.h>
```

Let's wrap it as:

```c
#ifdef CONFIG_MPU60X0_I2C
#  include <nuttx/sensors/mpu60x0.h>
#endif
```

The same for I2C:

```c
#ifdef CONFIG_I2C
#  include <nuttx/i2c/i2c_master.h>
#endif
```

**Comment by zouboan** [comment]
Thanks very much for your review @lupyuen, let me change the code!

File: drivers/sensors/mpu60x0.c
Comment (xiaoxiang781216) [inline drivers/sensors/mpu60x0.c:873]: snerr("ERROR: %p %p %zu\n", inode, dev, len);

File: drivers/sensors/mpu60x0.c
Comment (xiaoxiang781216) [inline drivers/sensors/mpu60x0.c:407]: it's better to change reg_addr type to uint8_t since mpu_regaddr_e may have 4bytes length and the cast only work for little endian machine.

Comment (zouboan) [reply drivers/sensors/mpu60x0.c]: done

File: drivers/sensors/mpu60x0.c
Comment (zouboan) [reply drivers/sensors/mpu60x0.c]: done

File: drivers/sensors/mpu60x0.c
Comment (zouboan) [reply drivers/sensors/mpu60x0.c]: done, thanks!

**Comment by acassis** [comment]
@zouboan I noticed this driver is not handling I2C_M_NOSTOP (and I2C_M_NOSTART). The best device to test it is the SSD1306 display because it uses the I2C_M_NOSTOP and will not work with not compatible I2C driver. I discovered that ESP32 suffer same issue and I'm preparing a patch to fix it.

File: arch/arm64/src/a64/a64_twi.c
Comment (zouboan) [inline arch/arm64/src/a64/a64_twi.c:1208]: @acassis here is handling of I2C_M_NOSTART

File: arch/arm64/src/a64/a64_twi.c
Comment (zouboan) [inline arch/arm64/src/a64/a64_twi.c:1214]: @acassis here is handling of I2C_M_NOSTART

File: arch/arm64/src/a64/a64_twi.c
Comment (zouboan) [inline arch/arm64/src/a64/a64_twi.c:1210]: @acassis here is handling of I2C_M_NOSTART

**Comment by zouboan** [comment]
> @zouboan I noticed this driver is not handling I2C_M_NOSTOP (and I2C_M_NOSTART). The best device to test it is the SSD1306 display because it uses the I2C_M_NOSTOP and will not work with not compatible I2C driver. I discovered that ESP32 suffer same issue and I'm preparing a patch to fix it.

mpu60x0 use both I2C_M_NOSTOP and I2C_M_NOSTART, see [https://github.com/apache/nuttx/blob/57e5d6b160d7dea1bda79f5102470a6c99156c5e/drivers/sensors/mpu60x0.c](url) L435~L440
I2C_M_NOSTART was handled in TWI driver, and I2C_M_NOSTOP was Implicit handled, because in function a64_twi_isr_process, there's no STOP when first I2C msg was transmited.
Test by Pinephone's MPU6050 , it seems the TWI driver worked fine.

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**
